File: app/src/main/java/org/softeg/slartus/forpdaplus/listfragments/next/UserReputationFragment.java
package org.softeg.slartus.forpdaplus.listfragments.next;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.Loader;
import android.view.ContextMenu;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.TextView;

import org.jetbrains.annotations.NotNull;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.softeg.slartus.forpdaapi.IListItem;
import org.softeg.slartus.forpdaapi.ListInfo;
import org.softeg.slartus.forpdaapi.ReputationEvent;
import org.softeg.slartus.forpdaapi.ReputationsApi;
import org.softeg.slartus.forpdaapi.classes.ListData;
import org.softeg.slartus.forpdaapi.classes.ReputationsListData;
import org.softeg.slartus.forpdaplus.App;
import org.softeg.slartus.forpdaplus.Client;
import org.softeg.slartus.forpdaplus.IntentActivity;
import org.softeg.slartus.forpdaplus.MainActivity;
import org.softeg.slartus.forpdaplus.R;
import org.softeg.slartus.forpdaplus.classes.ForumUser;
import org.softeg.slartus.forpdaplus.classes.MenuListDialog;
import org.softeg.slartus.forpdaplus.classes.common.ExtUrl;
import org.softeg.slartus.forpdaplus.fragments.profile.ProfileFragment;
import org.softeg.slartus.forpdaplus.listtemplates.UserReputationBrickInfo;
import org.softeg.slartus.forpdaplus.repositories.UserInfoRepository;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import okhttp3.Response;
import ru.slartus.http.Http;

/*
 * Created by slinkin on 19.02.2015.
 */
public class UserReputationFragment extends BrickFragmentListBase {
    public static final String USER_ID_KEY = "USER_ID_KEY";
    public static final String USER_NICK_KEY = "USER_NICK_KEY";
    public static final String USER_FROM_KEY = "USER_FROM_KEY";


    public static void showActivity(CharSequence userId, Boolean from) {
        Bundle args = new Bundle();
        args.putString(USER_ID_KEY, userId.toString());
        if (from)
            args.putBoolean(USER_FROM_KEY, true);
        MainActivity.showListFragment(userId.toString(), UserReputationBrickInfo.NAME, args);
    }

    @Override
    public boolean closeTab() {
        return false;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setArrow();
    }

    @Override
    public void onResume() {
        super.onResume();
        setArrow();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    private String getUserId() {
        return Args.getString(USER_ID_KEY);
    }

    private String getUserNick() {
        return Args.getString(USER_NICK_KEY, "");
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putString(USER_ID_KEY, getUserId());
        outState.putString(USER_NICK_KEY, getUserNick());
        super.onSaveInstanceState(outState);
    }

    @Override
    protected int getLoaderId() {
        return ItemsLoader.ID;
    }

    @Override
    protected BaseAdapter createAdapter() {
        return new ListAdapter(getActivity(), getData().getItems());
    }

    @Override
    protected int getViewResourceId() {
        return R.layout.list_fragment;
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        getActivity().openContextMenu(view);

    }

    @Override
    public void onLoadFinished(Loader<ListData> loader, ListData data) {
        super.onLoadFinished(loader, data);
        if (data.getEx() == null) {
            if (data instanceof ReputationsListData) {
                if (getSupportActionBar() != null)
                    setSubtitle(((ReputationsListData) data).getRep());
                Args.putString(USER_NICK_KEY, ((ReputationsListData) data).getUser());
            }
        }
    }

    @Override
    public View onCreateView(@NotNull android.view.LayoutInflater inflater, android.view.ViewGroup container,
                             Bundle savedInstanceState) {
        View v = super.onCreateView(inflater, container, savedInstanceState);
        addLoadMoreFooter(inflater.getContext());
        return v;
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) menuInfo;
        if (info.id == -1) return;
        final ReputationEvent item = (ReputationEvent) getAdapter().getItem((int) info.id);

        final List<MenuListDialog> list = new ArrayList<>();
        if (item.getSourceUrl() != null && !item.getSourceUrl().contains("forum/index.php?showuser=")) {
            list.add(new MenuListDialog(getString(R.string.jump_to_page), () -> IntentActivity.tryShowUrl(getActivity(), new Handler(), item.getSourceUrl(), true, false)));
        }
        ForumUser.onCreateContextMenu(getActivity(), list, item.getUserId(), item.getUser());
        ExtUrl.showContextDialog(getContext(), item.getUser(), list);
    }


    private static final String START_KEY = "START_KEY";

    protected Bundle getLoadArgs() {
        Bundle args = Args;
        args.putInt(START_KEY, getData().getItems().size());

        return args;
    }

    @Override
    public boolean onBackPressed() {
        return false;
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        return false;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        try {
            MenuItem item;
            if (Client.getInstance().getLogined() && !getUserId().equals(UserInfoRepository.Companion.getInstance().getId())) {


                item = menu.add(R.string.increase_reputation).setIcon(R.drawable.thumb_up);
                item.setOnMenuItemClickListener(menuItem -> {

                    plusRep();
                    return true;
                });
                item.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);

                item = menu.add(R.string.decrease_reputation).setIcon(R.drawable.thumb_down);
                item.setOnMenuItemClickListener(menuItem -> {
                    minusRep();
                    return true;
                });
                item.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
            }
            item = menu.add(R.string.Profile);
            item.setOnMenuItemClickListener(menuItem -> {
                ProfileFragment.showProfile(getUserId(), getUserNick());
                return true;
            });
            item.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
        } finally {
            super.onCreateOptionsMenu(menu, inflater);
        }
    }

    @Override
    protected AsyncTaskLoader<ListData> createLoader(int id, Bundle args) {
        ItemsLoader loader = null;
        if (id == ItemsLoader.ID) {


            setLoading(true);
            loader = new ItemsLoader(getActivity(), args);

        }
        return loader;
    }

    private static class ItemsLoader extends AsyncTaskLoader<ListData> {
        static final int ID = App.getInstance().getUniqueIntValue();
        ListData mApps;
        private Bundle args;


        ItemsLoader(Context context, Bundle args) {
            super(context);

            this.args = args;
        }

        public Bundle getArgs() {
            return args;
        }


        /**
         * Загрузка ссылки на изображение, которое является плюсовой репой
         */
        private void loadRepImage() throws IOException {
            String body = Client.getInstance().performGet("http://4pda.ru/forum/index.php?act=rep&view=history&mid=236113&mode=to&order=asc").getResponseBody();
            Element el = Jsoup
                    .parse(body)
                    .select("td.row1>img")
                    .first();
            if (el != null) {
                String plusImage = el.attr("src");
                if (plusImage != null) {
                    getPreferences().edit()
                            .putString("repPlusImage", plusImage)
                            .putBoolean("needLoadRepImage", false)
                            .apply();
                    return;
                }
            }
            if (el != null)
                el = el.select("tr:nth-last-child(2) td img").first();
            if (el != null) {
                String plusImage = el.attr("src");
                if (plusImage != null)
                    getPreferences().edit()
                            .putString("repPlusImage", plusImage)
                            .putBoolean("needLoadRepImage", false)
                            .apply();
            }
        }

        @Override
        public ListData loadInBackground() {
            try {
                loadRepImage();

                ListInfo listInfo = new ListInfo();
                listInfo.setFrom(getArgs().getBoolean(IS_REFRESH_KEY) ? 0 : getArgs().getInt(START_KEY));
                return ReputationsApi.loadReputation(Client.getInstance(),
                        args.getString(USER_ID_KEY),
                        args.getBoolean(USER_FROM_KEY), listInfo,
                        getPreferences().getString("repPlusImage", "http://s.4pda.to/ShmfPSURw3VD2aNlTerb3hvYwGCMxd4z0muJ.gif"));
            } catch (Throwable e) {
                ListData forumPage = new ListData();
                forumPage.setEx(e);

                return forumPage;
            }

        }

        @Override
        public void deliverResult(ListData apps) {

            mApps = apps;

            if (isStarted()) {
                super.deliverResult(apps);
            }

        }

        @Override
        protected void onStartLoading() {
            if (mApps != null) {
                // If we currently have a result available, deliver it
                // immediately.
                deliverResult(mApps);
            }

            if (takeContentChanged() || mApps == null) {
                // If the data has changed since the last time it was loaded
                // or is not currently available, start a load.
                forceLoad();
            }
        }


        @Override
        protected void onStopLoading() {
            // Attempt to cancel the current load task if possible.
            cancelLoad();
        }

        @Override
        protected void onReset() {
            super.onReset();

            // Ensure the loader is stopped
            onStopLoading();

            // At this point we can release the resources associated with 'apps'
            // if needed.
            if (mApps != null) {
                mApps = null;
            }
        }

    }

    private static class ListAdapter extends BaseAdapter {
        private final LayoutInflater mInflater;
        protected ArrayList<? extends IListItem> mData;


        ListAdapter(Context context, ArrayList<? extends IListItem> data) {
            mInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            mData = data;

        }

        public void setData(ArrayList<? extends IListItem> data) {
            mData = data;

        }

        @Override
        public int getCount() {
            return mData == null ? 0 : mData.size();
        }

        @Override
        public Object getItem(int i) {
            return mData.get(i);
        }

        @Override
        public long getItemId(int i) {
            return i;
        }

        @Override
        public android.view.View getView(int position, android.view.View view, android.view.ViewGroup parent) {
            final ViewHolder holder;
            if (view == null) {
                view = mInflater.inflate(R.layout.list_item_reputation, parent, false);
                holder = new ViewHolder();
                holder.Flag = view.findViewById(R.id.imgFlag);
                holder.TopLeft = view.findViewById(R.id.txtTopLeft);
                holder.TopRight = view.findViewById(R.id.txtTopRight);
                holder.Main = view.findViewById(R.id.txtMain);
                holder.SubMain = view.findViewById(R.id.txtSubMain);
                holder.progress = view.findViewById(R.id.progressBar);
                view.setTag(holder);

            } else {
                holder = (ViewHolder) view.getTag();
            }
            IListItem topic = mData.get(position);
            holder.TopLeft.setText(topic.getTopLeft());
            holder.TopRight.setText(topic.getTopRight());
            holder.Main.setText(topic.getMain());
            holder.SubMain.setText(topic.getSubMain());
            setVisibility(holder.progress, topic.isInProgress() ? View.VISIBLE : View.INVISIBLE);
            switch (topic.getState()) {
                case IListItem.STATE_GREEN:
                    setVisibility(holder.Flag, View.VISIBLE);
                    holder.Flag.setText("+");
                    holder.Flag.setBackgroundResource(R.drawable.plusrep);
                    //holder.Flag.setImageResource(R.drawable.new_flag);
                    break;
                case IListItem.STATE_RED:
                    setVisibility(holder.Flag, View.VISIBLE);
                    holder.Flag.setBackgroundResource(R.drawable.minusrep);
                    holder.Flag.setText("-");
                    //holder.Flag.setImageResource(R.drawable.old_flag);
                    break;
                default:
                    setVisibility(holder.Flag, View.INVISIBLE);
                    // holder.Flag.setImageBitmap(null);
            }
            return view;
        }

        private void setVisibility(View v, int visibility) {
            if (v.getVisibility() != visibility)
                v.setVisibility(visibility);
        }

        class ViewHolder {
            TextView Flag;
            View progress;
            TextView TopLeft;
            TextView TopRight;
            TextView Main;
            TextView SubMain;
        }
    }

    public void plusRep() {
        plusRep(getUserId(), getUserNick());
    }

    public void minusRep() {
        minusRep(getUserId(), getUserNick());
    }

    public void plusRep(String userId, String userNick) {
        plusRep(getActivity(), new Handler(), "0", userId, userNick);
    }

    public void minusRep(String userId, String userNick) {
        minusRep(getActivity(), new Handler(), "0", userId, userNick);
    }

    public static void plusRep(Activity activity, Handler handler, String userId, String userNick) {
        plusRep(activity, handler, "0", userId, userNick);
    }

    public static void minusRep(Activity activity, Handler handler, String userId, String userNick) {
        minusRep(activity, handler, "0", userId, userNick);
    }

    public static void plusRep(Activity activity, Handler handler, String postId, String userId, String userNick) {
        showChangeRep(activity, handler, postId, userId, userNick, "add", App.getContext().getString(R.string.increase_reputation));
    }

    public static void minusRep(Activity activity, Handler handler, String postId, String userId, String userNick) {
        showChangeRep(activity, handler, postId, userId, userNick, "minus", App.getContext().getString(R.string.decrease_reputation));
    }

    private static void showChangeRep(Activity activity, Handler handler, final String postId, String userId, String userNick, final String type, String title) {
        ForumUser.startChangeRep(activity, handler, userId, userNick, postId, type, title);
    }
}

File: app/src/main/java/org/softeg/slartus/forpdaplus/listfragments/next/UserReputationFragment.kt
package org.softeg.slartus.forpdaplus.listfragments.next

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.os.Handler
import androidx.loader.content.AsyncTaskLoader
import androidx.loader.content.Loader
import android.view.*
import android.widget.AdapterView
import android.widget.BaseAdapter
import android.widget.TextView
import org.jsoup.Jsoup
import org.softeg.slartus.forpdaapi.IListItem
import org.softeg.slartus.forpdaapi.ListInfo
import org.softeg.slartus.forpdaapi.ReputationEvent
import org.softeg.slartus.forpdaapi.ReputationsApi.loadReputation
import org.softeg.slartus.forpdaapi.classes.ListData
import org.softeg.slartus.forpdaapi.classes.ReputationsListData
import org.softeg.slartus.forpdaplus.*
import org.softeg.slartus.forpdaplus.classes.ForumUser
import org.softeg.slartus.forpdaplus.classes.MenuListDialog
import org.softeg.slartus.forpdaplus.classes.common.ExtUrl
import org.softeg.slartus.forpdaplus.fragments.GeneralFragment
import org.softeg.slartus.forpdaplus.fragments.profile.ProfileFragment
import org.softeg.slartus.forpdaplus.listtemplates.UserReputationBrickInfo
import org.softeg.slartus.forpdaplus.repositories.UserInfoRepositoryImpl.Companion.instance
import org.softeg.slartus.hosthelper.HostHelper
import java.io.IOException
import java.util.*

/*
 * Created by slinkin on 19.02.2015.
 */
class UserReputationFragment : BrickFragmentListBase() {
    override fun closeTab(): Boolean {
        return false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setArrow()
    }

    override fun onResume() {
        super.onResume()
        setArrow()
    }

    private val userId: String
        get() = Args.getString(USER_ID_KEY)?:""

    private val userNick: String
        get() = Args.getString(USER_NICK_KEY, "")

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString(USER_ID_KEY, userId)
        outState.putString(USER_NICK_KEY, userNick)
        super.onSaveInstanceState(outState)
    }

    override fun getLoaderId(): Int {
        return ItemsLoader.ID
    }

    override fun createAdapter(): BaseAdapter {
        return ListAdapter(activity, data.items)
    }

    override fun getViewResourceId(): Int {
        return R.layout.list_fragment
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
        activity!!.openContextMenu(view)
    }

    override fun onLoadFinished(loader: Loader<ListData>, data: ListData) {
        super.onLoadFinished(loader, data)
        if (data.ex == null) {
            if (data is ReputationsListData) {
                if (supportActionBar != null) setSubtitle(data.rep)
                Args.putString(USER_NICK_KEY, data.user)
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = super.onCreateView(inflater, container, savedInstanceState)
        addLoadMoreFooter(inflater.context)
        return v
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo?) {
        val info = menuInfo as AdapterView.AdapterContextMenuInfo
        if (info.id == -1L) return
        val item = adapter.getItem(info.id.toInt()) as ReputationEvent
        val list: MutableList<MenuListDialog> = ArrayList()
        if (item.sourceUrl != null && !item.sourceUrl.contains("forum/index.php?showuser=")) {
            list.add(MenuListDialog(getString(R.string.jump_to_page), Runnable { IntentActivity.tryShowUrl(activity, Handler(), item.sourceUrl, true, false) }))
        }
        ForumUser.onCreateContextMenu(activity, list, item.userId, item.user)
        ExtUrl.showContextDialog(context, item.user, list)
    }

    override fun getLoadArgs(): Bundle {
        val args = Args
        args.putInt(START_KEY, data.items.size)
        return args
    }

    override fun onBackPressed(): Boolean {
        return false
    }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        return false
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.reputation, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.profile_rep_item -> {
                ProfileFragment.showProfile(userId, userNick)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        val loginedAndNotSelf = Client.getInstance().logined && userId != instance.getId()
    }

    override fun createLoader(id: Int, args: Bundle): AsyncTaskLoader<ListData> {
        var loader: ItemsLoader? = null
        if (id == ItemsLoader.ID) {
            setLoading(true)
            loader = ItemsLoader(activity, args)
        }
        return loader!!
    }

    private class ItemsLoader(context: Context?, val args: Bundle) : AsyncTaskLoader<ListData>(context!!) {
        var mApps: ListData? = null

        /**
         * Загрузка ссылки на изображение, которое является плюсовой репой
         */
        @Throws(IOException::class)
        private fun loadRepImage() {
            val body = Client.getInstance().performGet("https://${HostHelper.host}/forum/index.php?act=rep&view=history&mid=236113&mode=to&order=asc").responseBody
            var el = Jsoup
                    .parse(body)
                    .select("td.row1>img")
                    .first()
            if (el != null) {
                val plusImage = el.attr("src")
                if (plusImage != null) {
                    GeneralFragment.getPreferences().edit()
                            .putString("repPlusImage", plusImage)
                            .putBoolean("needLoadRepImage", false)
                            .apply()
                    return
                }
            }
            if (el != null) el = el.select("tr:nth-last-child(2) td img").first()
            if (el != null) {
                val plusImage = el.attr("src")
                if (plusImage != null) GeneralFragment.getPreferences().edit()
                        .putString("repPlusImage", plusImage)
                        .putBoolean("needLoadRepImage", false)
                        .apply()
            }
        }

        override fun loadInBackground(): ListData {
            return try {
                loadRepImage()
                val listInfo = ListInfo()
                listInfo.from = if (args.getBoolean(IS_REFRESH_KEY)) 0 else args.getInt(START_KEY)
                loadReputation(Client.getInstance(),
                        args.getString(USER_ID_KEY),
                        args.getBoolean(USER_FROM_KEY), listInfo,
                        GeneralFragment.getPreferences().getString("repPlusImage", "https://s.4pda.to/ShmfPSURw3VD2aNlTerb3hvYwGCMxd4z0muJ.gif"))
            } catch (e: Throwable) {
                val forumPage = ListData()
                forumPage.ex = e
                forumPage
            }
        }

        override fun deliverResult(apps: ListData?) {
            mApps = apps
            if (isStarted) {
                super.deliverResult(apps)
            }
        }

        override fun onStartLoading() {
            if (mApps != null) { // If we currently have a result available, deliver it
// immediately.
                deliverResult(mApps)
            }
            if (takeContentChanged() || mApps == null) { // If the data has changed since the last time it was loaded
// or is not currently available, start a load.
                forceLoad()
            }
        }

        override fun onStopLoading() { // Attempt to cancel the current load task if possible.
            cancelLoad()
        }

        override fun onReset() {
            super.onReset()
            // Ensure the loader is stopped
            onStopLoading()
            // At this point we can release the resources associated with 'apps'
// if needed.
            if (mApps != null) {
                mApps = null
            }
        }

        companion object {
            val ID = App.getInstance().uniqueIntValue
        }

    }

    private class ListAdapter(context: Context?, data: ArrayList<out IListItem>) : BaseAdapter() {
        private val mInflater: LayoutInflater = context!!.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        private var mData: ArrayList<out IListItem> = data


        override fun getCount(): Int {
            return mData.size
        }

        override fun getItem(i: Int): Any {
            return mData[i]
        }

        override fun getItemId(i: Int): Long {
            return i.toLong()
        }

        override fun getView(position: Int, v: View?, parent: ViewGroup): View? {
            var view = v
            val holder: ViewHolder
            if (view == null) {
                view = mInflater.inflate(R.layout.list_item_reputation, parent, false)
                holder = ViewHolder()
                holder.flag = view.findViewById(R.id.imgFlag)
                holder.topLeft = view.findViewById(R.id.txtTopLeft)
                holder.topRight = view.findViewById(R.id.txtTopRight)
                holder.main = view.findViewById(R.id.txtMain)
                holder.subMain = view.findViewById(R.id.txtSubMain)
                holder.progress = view.findViewById(R.id.progressBar)
                view.tag = holder
            } else {
                holder = view.tag as ViewHolder
            }
            val topic = mData[position]
            holder.topLeft!!.text = topic.topLeft
            holder.topRight!!.text = topic.topRight
            holder.main!!.text = topic.main
            holder.subMain!!.text = topic.subMain
            setVisibility(holder.progress, if (topic.isInProgress) View.VISIBLE else View.INVISIBLE)
            when (topic.state) {
                IListItem.STATE_GREEN -> {
                    setVisibility(holder.flag, View.VISIBLE)
                    holder.flag!!.text = "+"
                    holder.flag!!.setBackgroundResource(R.drawable.plusrep)
                }
                IListItem.STATE_RED -> {
                    setVisibility(holder.flag, View.VISIBLE)
                    holder.flag!!.setBackgroundResource(R.drawable.minusrep)
                    holder.flag!!.text = "-"
                }
                else -> setVisibility(holder.flag, View.INVISIBLE)
            }
            return view
        }

        private fun setVisibility(v: View?, visibility: Int) {
            if (v!!.visibility != visibility) v.visibility = visibility
        }

        inner class ViewHolder {
            var flag: TextView? = null
            var progress: View? = null
            var topLeft: TextView? = null
            var topRight: TextView? = null
            var main: TextView? = null
            var subMain: TextView? = null
        }

    }

    @JvmOverloads
    fun plusRep(uId: String? = userId, uNick: String? = userNick) {
        plusRep(activity, Handler(), "0", uId, uNick)
    }

    @JvmOverloads
    fun minusRep(uId: String? = userId, uNick: String? = userNick) {
        minusRep(activity, Handler(), "0", uId, uNick)
    }

    companion object {
        const val USER_ID_KEY = "USER_ID_KEY"
        const val USER_NICK_KEY = "USER_NICK_KEY"
        const val USER_FROM_KEY = "USER_FROM_KEY"
        @JvmStatic
        fun showActivity(userId: CharSequence, from: Boolean) {
            val args = Bundle()
            args.putString(USER_ID_KEY, userId.toString())
            if (from) args.putBoolean(USER_FROM_KEY, true)
            MainActivity.showListFragment(userId.toString(), UserReputationBrickInfo.NAME, args)
        }

        private const val START_KEY = "START_KEY"

        @JvmStatic
        fun plusRep(activity: Activity?, handler: Handler?, userId: String?, userNick: String?) {
            plusRep(activity, handler, "0", userId, userNick)
        }

        @JvmStatic
        fun minusRep(activity: Activity?, handler: Handler?, userId: String?, userNick: String?) {
            minusRep(activity, handler, "0", userId, userNick)
        }

        @JvmStatic
        fun plusRep(activity: Activity?, handler: Handler?, postId: String, userId: String?, userNick: String?) {
            showChangeRep(activity, handler, postId, userId, userNick, "add", App.getContext().getString(R.string.increase_reputation))
        }

        @JvmStatic
        fun minusRep(activity: Activity?, handler: Handler?, postId: String, userId: String?, userNick: String?) {
            showChangeRep(activity, handler, postId, userId, userNick, "minus", App.getContext().getString(R.string.decrease_reputation))
        }

        private fun showChangeRep(activity: Activity?, handler: Handler?, postId: String, userId: String?, userNick: String?, type: String, title: String) {
            ForumUser.startChangeRep(activity, handler, userId, userNick, postId, type, title)
        }
    }
}